mne/source_tfr.py

```diff
@@ -0,0 +1,305 @@
+        %(verbose_meth)s
+
+        Notes
+        -----
+        For some data, it may be more accurate to use npad=0 to reduce
+        artifacts. This is dataset dependent -- check your data!
+
+        Note that the sample rate of the original data is inferred from tstep.
+        """
+        # resampling in sensor instead of source space gives a somewhat
+        # different result, so we don't allow it
+        self._remove_kernel_sens_data_()
+
+        o_sfreq = 1.0 / self.tstep
+        self.data = resample(self.data, sfreq, o_sfreq, npad, n_jobs=n_jobs)
+
+        # adjust indirectly affected variables
+        self.tstep = 1.0 / sfreq
+        return self
+
+    @property
+    def data(self):
+        """Numpy array of SourceTFR data."""
+        if self._data is None:
+            # compute the solution the first time the data is accessed and
+            # remove the kernel and sensor data
+            self._remove_kernel_sens_data_()
+        return self._data
+
+    @data.setter
+    def data(self, value):
+        value = np.asarray(value)
+        if self._data is not None and value.ndim != self._data.ndim:
+            raise ValueError('Data array should have %d dimensions.' %
+                             self._data.ndim)
+
+        # vertices can be a single number, so cast to ndarray
+        if isinstance(self.vertices, list):
+            n_verts = sum([len(v) for v in self.vertices])
+        elif isinstance(self.vertices, np.ndarray):
+            n_verts = len(self.vertices)
+        else:
+            raise ValueError('Vertices must be a list or numpy array')
+
+        if value.shape[0] != n_verts:
+            raise ValueError('The first dimension of the data array must '
+                             'match the number of vertices (%d != %d)' %
+                             (value.shape[0], n_verts))
+
+        self._data = value
+        self._update_times()
+
+    @property
+    def shape(self):
+        """Shape of the data."""
+        if self._data is not None:
+            return self._data.shape
+        return (self._kernel.shape[0],
+                self._sens_data.shape[1],
+                self._sens_data.shape[2])
+
+    @property
+    def tmin(self):
+        """The first timestamp."""
+        return self._tmin
+
+    @tmin.setter
+    def tmin(self, value):
+        self._tmin = float(value)
+        self._update_times()
+
+    @property
+    def tstep(self):
+        """The change in time between two consecutive samples (1 / sfreq)."""
+        return self._tstep
+
+    @tstep.setter
+    def tstep(self, value):
+        if value <= 0:
+            raise ValueError('.tstep must be greater than 0.')
+        self._tstep = float(value)
+        self._update_times()
+
+    @property
+    def times(self):
+        """A timestamp for each sample."""
+        return self._times
+
+    @times.setter
+    def times(self, value):
+        raise ValueError('You cannot write to the .times attribute directly. '
```

Comment:
why is this a value error? isn't it a runtime?

Is there any other way to prevent the setter? maybe not.